pq/autocomplete/earley/earley.go

```diff
@@ -133,61 +133,60 @@ func (p *Earley) resizeChart(size int) {
 // grammar, which simplifies our parsing logic considerably.
 func (p *Earley) Parse(input string) *earleyChart {
 	inputWords := extractWords(input)
-	p.chart.setInputWords(inputWords)
 	return p.ParseTokens(inputWords)
 }
 
-// Parse parses the full input string. It first tokenizes the input
-// string and then uses those tokens as atomic units in our
-// grammar, which simplifies our parsing logic considerably.
+// ParseTokens parses the full input tokens from beginning
 func (p *Earley) ParseTokens(tokens Tokens) *earleyChart {
-	p.words = tokens
-	p.resizeChartIfNecessary(tokens)
-	// parse through all words, the last of the token is always Eof
-	for stateIndex := 0; stateIndex < len(tokens); stateIndex++ {
-		token := tokens[stateIndex]
-		p.PartialParse(token, stateIndex)
-		debug.Debugf("------\n%v\n------\n", p.chart.String())
-	}
+	p.chart.resetChart()
+	p.PartialParse(tokens, 0)
+	debug.Debugf("------\n%v\n------\n", p.chart.String())
 	return p.chart
 }
 
 // This is the incremental bit of our parser. We can basically feed in
-// the next token and parse at a specific word index.
-func (p *Earley) PartialParse(token Tokhan, chartIndex int) *StateSet {
+// a list of words (i.e. lexed tokens) and parse at a specific word index.
+func (p *Earley) PartialParse(tokens Tokens, chartIndex int) *earleyChart {
 	debug.Debugf("------- starting partial parse %v\n", chartIndex)
-	//p.words = tokens
 	// we're going to assume here that we've correctly parsed prior to our index
-	//p.resizeChartIfNecessary(tokens)
+	p.words = append(p.words, tokens...)
+	p.resizeChartIfNecessary(p.words)
+	p.chart.setInputWords(p.words)
 	currStateSet := p.chart.GetState(chartIndex)
-	setIndex := 0
-	for {
-		if setIndex >= len(currStateSet.GetStates()) {
-			break
-		}
-		item := currStateSet.items[setIndex]
-		fromItems := []ItemId{item.id}
-		if !item.isCompleted() {
-			// predict if current state isn't terminal
-			if !item.GetRightSymbolByRulePos().isTerminal() {
-				p.predict(item, chartIndex, fromItems)
-			} else {
-				// Scan the next symbol which is terminal
-				p.scan(item, chartIndex, fromItems, token)
+	for _, token := range tokens {
+		setIndex := 0
+		for {
+			if setIndex >= len(currStateSet.GetStates()) {
+				break
+			}
+			item := currStateSet.items[setIndex]
+			fromItems := []ItemId{item.id}
+			if !item.isCompleted() {
+				// predict if current state isn't terminal
+				if !item.GetRightSymbolByRulePos().isTerminal() {
+					p.predict(item, chartIndex, fromItems)
+				} else {
+					// Scan the next symbol which is terminal
+					p.scan(item, chartIndex, fromItems, token)
+				}
+			} else { // end of rule, let's complete
+				p.complete(item, chartIndex)
 			}
-		} else { // end of rule, let's complete
-			p.complete(item, chartIndex)
+			setIndex++
 		}
-		setIndex++
+		chartIndex++
+		currStateSet = p.chart.GetState(chartIndex)
+		debug.Debugf("------\n%v\n------\n", p.chart.String())
 	}
-	return currStateSet
+	return p.chart
```

Comment:
I don't understand this return anymore.. if we call p.PartialParse(tokens,index) -> shouldn't we expect to return the stateset at index?

Reply:
My understanding is that `ParseToken` and `PartialParse` basically do the same thing. The difference is that `PartialParse` parses from the input index and `ParseTokens` parses from the beginning. `ParseTokens` calls `PartialParse` with parameter chartIndex 0. If so, then both of them should return chart. Probably my understanding is different from your original thoughts? 

I added the usage of PartialParse in function `GetSuggestedTokenType` and also added the testcases.